cmd/auth-provider-gcp/app/getcredentials_test.go

```diff
@@ -0,0 +1,118 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"errors"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func TestValidateAuthFlow(t *testing.T) {
+	type FlagResult struct {
+		Name  string
+		Flow  string
+		Error error
+	}
+	tests := []FlagResult{
+		{Name: "validate gcr auth flow", Flow: gcrAuthFlow},
+		{Name: "validate docker-cfg auth flow option", Flow: dockerConfigAuthFlow},
+		{Name: "validate docker-cfg-url auth flow option", Flow: dockerConfigURLAuthFlow},
+		{Name: "bad auth flow option", Flow: "bad-flow", Error: &AuthFlowFlagError{flagValue: "bad-flow"}},
+		{Name: "empty auth flow option", Flow: "", Error: &AuthFlowFlagError{flagValue: ""}},
+		{Name: "case-sensitive auth flow", Flow: "Gcrauthflow", Error: &AuthFlowFlagError{flagValue: "Gcrauthflow"}},
+	}
+	for _, tc := range tests {
+		t.Run(tc.Name, func(t *testing.T) {
+			err := validateFlags(&CredentialOptions{AuthFlow: tc.Flow})
+			if tc.Error != nil {
+				if err == nil {
+					t.Fatalf("with flow %q did not get expected error %q", tc.Flow, err)
+				}
+				if !errors.Is(err, tc.Error) {
+					t.Fatalf("with flow %q got unexpected error type %q (expected %q)", tc.Flow, reflect.TypeOf(err), reflect.TypeOf(tc.Error))
+				}
+				return
+			}
+			if err != nil {
+				t.Fatalf("with flow %q unexpected error %q", tc.Flow, err)
+			}
+		})
+	}
+}
+
+func TestProviderFromFlow(t *testing.T) {
+	type ProviderResult struct {
+		Name  string
+		Flow  string
+		Type  string
+		Error error
+	}
+	tests := []ProviderResult{
+		{Name: "gcr auth provider selection", Flow: gcrAuthFlow, Type: "ContainerRegistryProvider"},
+		{Name: "docker-cfg auth provider selection", Flow: dockerConfigAuthFlow, Type: "DockerConfigKeyProvider"},
+		{Name: "docker-cfg-url auth provider selection", Flow: dockerConfigURLAuthFlow, Type: "DockerConfigURLKeyProvider"},
+		{Name: "non-existent auth provider request", Flow: "bad-flow", Type: "", Error: &AuthFlowTypeError{requestedFlow: "bad-flow"}},
+		{Name: "empty auth provider request", Flow: "", Type: "", Error: &AuthFlowTypeError{requestedFlow: ""}},
+	}
+	for _, tc := range tests {
+		t.Run(tc.Name, func(t *testing.T) {
+			provider, err := providerFromFlow(tc.Flow)
+			if tc.Error != nil {
+				if err == nil {
+					t.Fatalf("with flow %q did not get expected error %q", tc.Flow, err)
+				}
+				if !errors.Is(err, tc.Error) {
+					t.Fatalf("with flow %q got unexpected error type %q (expected %q)", tc.Flow, reflect.TypeOf(err), reflect.TypeOf(tc.Error))
+				}
+				return
+			}
+			if err != nil {
+				t.Fatalf("with flow %q unexpected error %q", tc.Flow, err)
+			}
+			providerType := reflect.TypeOf(provider).String()
+			if providerType != "*gcpcredential."+tc.Type {
+				t.Errorf("with flow %q unexpected provider type %q", tc.Flow, providerType)
+			}
+		})
+	}
+}
+
+func TestFlagError(t *testing.T) {
+	badFlow := "bad-flow"
```

Comment:
There is nothing wrong with doing table style testing on a table of length 1. It makes it easier on the devs who come later. You have 1 consistent style. In addition if someone wanted to add something like a case sensitivity test or added a second flag and wanted an error test for it, it would be easier to fold that in.

Reply:
Gotcha, done